### DISCUSSION TITLE
Add nullable support

### DISCUSSION THREAD

**Discussion by Panthr75** [discord-net/Discord.Net#2697] [Ideas]
Include `<Nullable>enable</Nullable>` into csproj. I hate shooting myself in the foot due to the lack of nullable annotations.

First, take a notice of SocketGuild.GetTextChannel(ulong):
```cs
// SocketGuild.GetTextChannel.cs
namespace Discord.WebSocket;
public partial class SocketGuild
{
  /// <summary>
  /// Gets a text channel in this guild
  /// </summary>
  /// <param name="id">The snowflake identifier for the text channel.</param>
  /// <returns>A text channel associated with the specified <paramref name="id"/>; <see langword="null"/> if none is found.</returns>
  public SocketTextChannel GetTextChannel(ulong id);
}
```

There are no nullable annotations dictating SocketTextChannel could be null without explicitly reading the XML documentation. This can lead to this:
```cs
// basic greeting stubbed from Little Devil
DiscordClient client = new();
//...

client.OnUserJoinedGuild = async (SocketGuildUser user) =>
{
  string greetingMessage;
  ulong greetingChannelId;
  // fetch greeting message and greeting channel id from database using user.Guild.Id
  
  SocketTextChannel textChannel = user.Guild.GetTextChannel(greetingChannelId);
  // RuntimeError: NullReferenceException
  await textChannel.SendMessageAsync(greetingMessage); 
};
```

This code works when testing, due to the channel ID properly being found. But when this code is pushed to production, NullReferenceExceptions occur due to admins deleting guild channels, and thus the text channel wont be found.
By updating the library with proper nullable annotations, there are a lot less chances to shoot yourself in the foot:
```cs
// SocketGuild.GetTextChannel.cs
namespace Discord.WebSocket;
public partial class SocketGuild
{
  /// <summary>
  /// Gets a text channel in this guild
  /// </summary>
  /// <param name="id">The snowflake identifier for the text channel.</param>
  /// <returns>A text channel associated with the specified <paramref name="id"/>; <see langword="null"/> if none is found.</returns>
  public SocketTextChannel? GetTextChannel(ulong id); // Added a ? to SocketTextChannel, to indicate it could be null
}
```

```cs
// basic greeting stubbed from Little Devil
DiscordClient client = new();
//...

client.OnUserJoinedGuild = async (SocketGuildUser user) =>
{
  string greetingMessage;
  ulong greetingChannelId;
  // fetch greeting message and greeting channel id from database using user.Guild.Id
  
  // Compile Time Warning: CS8600 - Converting null literal or possible null value to non-nullable type. 
  SocketTextChannel textChannel = user.Guild.GetTextChannel(greetingChannelId);
  // RuntimeError: NullReferenceException
  await textChannel.SendMessageAsync(greetingMessage); 
};
```

Doing this would also make clear whether or not something COULD be null at any point, like AvatarURL or something of the sorts.

Especially with #2696, it's crucial to indicate that `GlobalName` is `nullable`, or else suddenly your code can experience a NullReferenceException as you didn't directly read the XML documentation.

The only caveats would be not being able to compile for .NET Framework 4.6.1, and .NET Standard 2.0, but both are out of date, with .NET Framework 4.6.1 being released in 2015 (8 years from now), and .NET Standard 2.0 being released in mid 2017 (almost 6 years from now). .NET generates the necessary attribute classes as internal, so packages referencing the code without nullable wont be affected at all.

Overall, adding nullability increases the productivity of developers using Discord.NET, and prevents newer C# devs and even new devs using the library from putting in code that could at one point `NullReferenceException`.

**Comment by Misha-133** [comment]
Dropping support for .NET framework would be a major breaking change. We have plans to drop it in v4 tho. And also we plan on adding nullable support in v4 too.
